lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+                [Service]
+                Type=simple
+                Restart=always
+                WorkingDirectory={Path.cwd()}
+                Environment="PYTHONPATH={Path.cwd() / "venv"}"
+                ExecStart=/usr/bin/python3 {__file__} {self._charm.unit.name}
+
+                [Install]
+                WantedBy=multi-user.target
+                """
+            ).strip()
+        )
+        _logger.debug(f"Service file {self._service_file.name} written. Reloading systemd")
+        _daemon_reload()
+        # Notices daemon is enabled so that the service will start even after machine reboot.
+        # This functionality is needed in the event that a charm is rebooted to apply updates.
+        _enable_service(self._service_file.name)
+        _start_service(self._service_file.name)
+        _logger.debug(f"Started {self._service_file.name} daemon")
+
+    def stop(self) -> None:
+        """Stop charmed operator from observing the status of subscribed services."""
+        _stop_service(self._service_file.name)
+        # Notices daemon is disabled so that the service will not restart after machine reboot.
+        _disable_service(self._service_file.name)
+
+
+def _name_to_dbus_path(name: str) -> str:
+    """Convert the specified name into an org.freedesktop.systemd1.Unit path handle.
+
+    Args:
+        name: The name of the service.
+
+    Returns:
+        String containing the DBus path.
+    """
+    # DBUS Object names may only contain ASCII chars [A-Z][a-z][0-9]_
+    # It's basically urlencoded but instead of a %, it uses a _
+    path = name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        path = path.replace(value, key)
+
+    return f"/org/freedesktop/systemd1/unit/{path}"
+
+
+def _dbus_path_to_name(path: str) -> str:
+    """Convert the specified DBus path handle to a service name.
+
+    Args:
+        path: The DBus path to convert to service name.
+
+    Returns:
+        String containing the service name.
+    """
+    # DBUS Object names may only contain ASCII chars [A-Z][a-z][0-9]_
+    name = Path(path).name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        name = name.replace(key, value)
+
+    return name
+
+
+def _systemd_unit_changed(msg: Message) -> bool:
+    """Send Juju notification if systemd unit state changes on the DBus bus.
+
+    Invoked when a PropertiesChanged event occurs on an org.freedesktop.systemd1.Unit
+    object across the dbus. These events are sent whenever a unit changes state, including
+    starting and stopping.
+
+    Args:
+        msg: The message to process in the callback.
+
+    Returns:
+        True if the event is processed. False if otherwise.
+    """
+    _logger.debug(
+        f"Received message: path: {msg.path}, interface: {msg.interface}, member: {msg.member}"
+    )
+    service = _dbus_path_to_name(msg.path)
+    properties = msg.body[1]
+    if "ActiveState" not in properties:
+        return False
+
+    global _service_states
```

Comment:
Are there any concurrency issues with global state being used/updated in this way? Or is it sequential access due to how async works? I'm not super familiar with it, but I presume it runs everything on a single-threaded loop.

Reply:
I haven't seen any issues since we don't create any separate threads or event loops. We did initially have a ThreadPoolExecutor to get around a RuntimeError we were encountering when trying to call the async event loop when there was none (synchronous -> asynchronous), but we fixed that issue by early binding the event loop to the lambda that invokes the synchronous `_load_services(...)` function.